src/Solver.cpp

```diff
@@ -315,16 +391,90 @@ void Solver::setPolarQuadrature(PolarQuad* polar_quad) {
 
 
 /**
- * @brief Sets the threshold for source convergence (>0)
- * @param source_thresh the threshold for source convergence
+ * @brief Sets the threshold for source/flux convergence.
+ * @brief The default threshold for convergence is 1E-5.
+ * @param source_thresh the threshold for source/flux convergence
+ */
+void Solver::setConvergenceThreshold(FP_PRECISION threshold) {
+
+  if (threshold <= 0.0)
+    log_printf(ERROR, "Unable to set the convergence threshold to %f "
+               "since it is not a positive number", threshold);
+
+  _converge_thresh = threshold;
+}
+
+
+/**
+ * @brief Assign a fixed source for a flat source region and energy group.
+ * @details This is a helper routine to perform error checking for the
+ *          subclasses which store the source in the appropriate array.
+ * @param fsr_id the flat source region ID
+ * @param group the energy group
+ * @param source the volume-averaged source in this group
+ */
+void Solver::setFixedSourceByFSR(int fsr_id, int group, FP_PRECISION source) {
+  
+  if (group <= 0 || group > _num_groups)
+    log_printf(ERROR,"Unable to set fixed source for group %d in "
+               "in a %d energy group problem", group, _num_groups);
+
+  if (fsr_id < 0 || fsr_id >= _num_FSRs)
+    log_printf(ERROR,"Unable to set fixed source for FSR %d with only "
+               "%d FSRs in the geometry", fsr_id, _num_FSRs);
+}
+
+
+/**
+ * @brief Assign a fixed source for a Cell and energy group.
+ * @details This routine will add the fixed source to all instances of the
+ *          Cell in the geometry (e.g., all FSRs for this Cell).
+ * @param fsr_id the Cell of interest
+ * @param group the energy group
+ * @param source the volume-averaged source in this group
+ */
+void Solver::setFixedSourceByCell(Cell* cell, int group, FP_PRECISION source) {
+
+  /* Recursively add the source to all Cells within a FILL type Cell */
+  if (cell->getType() == FILL) {
+    std::map<int, Cell*> cells = cell->getAllCells();
+    std::map<int, Cell*>::iterator iter;
+    for (iter = cells.begin(); iter != cells.end(); ++iter)
+      setFixedSourceByCell(iter->second, group, source);
+  }
+
+  /* Aadd the source to all FSRs for this MATERIAL type Cell */
+  else {
+    CellBasic* fsr_cell;
+    
+    for (int r=0; r < _num_FSRs; r++) {
+      fsr_cell = _geometry->findCellContainingFSR(r);
+      if (cell->getId() == fsr_cell->getId())
+        setFixedSourceByFSR(r, group, source);
+    }
+  }
+}
+
```

Comment:
Yes, this is a tricky one. The reason that this occurs is that the
``Solver``  is blind to the ``Cell`` "clones" that are created for each
ring/sector during the ``Geometry::initializeFlatSourceRegions()`` routine.
The original "parent" ``Cell``  is kept as one ring/sector, but all other
``Cell`` clones have new IDs which correspond to different FSRs. Hence,
when you set the source for a particular ``Cell``, the ``Solver`` will only
register that source for that ``Cell`` and not all of its clones. This
leaves it to the user to figure out how to register sources for each
ring/sector which is a pain but can be done with some bookkeeping.

Perhaps one good option for resolving this would be some routines like
``Solver::setFixedSource(char* cell_name, FP_PRECISION source)``. This
would permit the user to leverage ``Cell`` string names which are inherited
by the clones. What do you think @PrezNattyGibbs?
On May 20, 2015 6:53 PM, "Nathan A. Gibson" <notifications@github.com>
wrote:

> In src/Solver.cpp
> <https://github.com/mit-crpg/OpenMOC/pull/142#discussion_r30758002>:
>
> > +    for (iter = cells.begin(); iter != cells.end(); ++iter)
> > +      setFixedSourceByCell(iter->second, group, source);
> > +  }
> > +
> > +  /* Aadd the source to all FSRs for this MATERIAL type Cell */
> > +  else {
> > +    CellBasic* fsr_cell;
> > +
> > +    for (int r=0; r < _num_FSRs; r++) {
> > +      fsr_cell = _geometry->findCellContainingFSR(r);
> > +      if (cell->getId() == fsr_cell->getId())
> > +        setFixedSourceByFSR(r, group, source);
> > +    }
> > +  }
> > +}
> > +
>
> I just tried this for a cell with rings, and it looks like the source only
> made it into the outer ring. I'm not really sure what went wrong. I'll look
> into it more later if you don't first.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/OpenMOC/pull/142/files#r30758002>.
>